executor/src/witgen/processor.rs

```diff
@@ -148,27 +159,44 @@ impl<'a, 'b, T: FieldElement, Q: QueryCallback<T>, CalldataAvailable>
     ) -> Result<Constraints<&'a PolynomialReference, T>, EvalError<T>> {
         let mut outer_assignments = vec![];
 
-        while let Some(SequenceStep {
-            row_delta,
-            identity,
-        }) = sequence_iterator.next()
-        {
+        while let Some(SequenceStep { row_delta, action }) = sequence_iterator.next() {
             let row_index = (1 + row_delta) as usize;
-            let progress = match identity {
-                IdentityInSequence::Internal(identity_index) => {
+            let progress = match action {
+                Action::InternalIdentity(identity_index) => {
                     self.process_identity(row_index, identity_index)?
                 }
-                IdentityInSequence::OuterQuery => {
+                Action::OuterQuery => {
                     let (progress, new_outer_assignments) = self.process_outer_query(row_index)?;
                     outer_assignments.extend(new_outer_assignments);
                     progress
                 }
+                Action::ProverQueries => self.process_queries(row_index),
             };
             sequence_iterator.report_progress(progress);
         }
         Ok(outer_assignments)
     }
 
+    fn process_queries(&mut self, row_index: usize) -> bool {
+        let mut query_processor =
+            QueryProcessor::new(self.fixed_data, self.mutable_state.query_callback);
+        let global_row_index = self.row_offset + row_index as u64;
+        let row_pair = RowPair::new(
+            &self.data[row_index],
+            &self.data[row_index + 1],
+            global_row_index,
+            self.fixed_data,
+            UnknownStrategy::Unknown,
+        );
+        let mut updates = EvalValue::complete(vec![]);
+        for poly_id in self.fixed_data.witness_cols.keys() {
+            if self.is_relevant_witness[&poly_id] {
+                updates.combine(query_processor.process_query(&row_pair, &poly_id));
+            }
+        }
+        self.apply_updates(row_index, &updates, || "queries".to_string())
```

Comment:
Hm, this is the only part that needs a mutable ref to self, right? So you could extract building the row pair into a function if you also extract all the stuff above here into a function, right?